### PR TITLE
Basic implementation of geometry optimization interface.

### PR DESCRIPTION
**Pull request by CedricTravelletti**
This is a basic implementation of the geometry optimization interface. It allows for plug-and-play modifications of the computational backend through the specification of an AtomsCalculators and also for using various optimization libraries via Optimization.jl. 

It features a bunch of examples showcasing the above capabilities.

### CODE REVIEW

**Comment by cortner** [comment]
Thanks, Cedric! @tjjarvinen  please do a first-round review and I'll then look at it also as soon as I can manage, probably within the next 2-3 days.

**Comment by cortner** [comment]
We can add a Stillinger-Weber and a Lennard-Jones test either to the examples or to the tests even.

File: src/atomsbase_interface.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+@doc raw"""
+    update_postions(system::AbstractSystem, positions::Vector{<:AbstractVector{<:Real}}) where {L <: Unitful.Length}
+
+Creates a new system based on ``system`` but with atoms positions updated to the ones specified in `positions`, 
+using cartesian coordinates.
+"""
+function update_positions(system::AbstractSystem, positions::AbstractVector{<:AbstractVector{L}}) where {L <: Unitful.Length}
+    particles = [
+                 Atom(atom, position=SVector{3,L}(position)) 
+                 for (atom, position) in zip(system.particles, positions)]
```
Comment (tjjarvinen) [inline src/atomsbase_interface.jl:23]: This is a problematic

```julia
julia> FlexibleSystem <: AbstractSystem
true

julia> FastSystem <: AbstractSystem
true
```

But `FastSystem` does not have `.particles`. AtomsBase does define iterator over particles, so you can just do

```julia
for (atom, position) in zip(system, positions)]
```
and it will work.

With this you can change

```julia
unction update_positions(system::AbstractSystem, positions::AbstractVector{<:AbstractVector{L}}) where {L <: Unitful.Length}
```

to 

```julia
unction update_positions(system, positions::AbstractVector{<:AbstractVector{L}}) where {L <: Unitful.Length}
```

Also `AbstractSystem` is actually an abstract type and calls like `AbstractSystem(system, particles=particles)` do call concrete types, in this case `FlexibleSystem`. It might be more clear to just call it directly.

File: src/atomsbase_interface.jl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+@doc raw"""
+    update_optimizable_coordinates(system::AbstractSystem, positions::Vector{<:AbstractVector{<:L}}) where {L <: Unitful.Length}
+
+Creates a new system based on ``system`` with the optimizable coordinates are
+updated to the ones provided (in the order in which they appear in system.particles), cartesian coordinates version..
+"""
+function update_optimizable_coordinates(system::AbstractSystem, positions::AbstractVector{<:L}) where {L <: Unitful.Length}
+    mask = get_optimizable_mask(system)
+    new_positions = Vector.(position(system)) # make mutable.
+    new_positions[mask] = [eachcol(reshape(positions, 3, :))...]
```
Comment (tjjarvinen) [inline src/atomsbase_interface.jl:36]: You could use `reinterpret` here to remove allocation

```julia
reinterpret(reshape, SVector{3, eltype(positions)}, reshape(positions, 3, :))
```

Also when allocating similar array you should use `similar`. So, `similar(position(system))` instead of `Vector.(position(system))`. Or, if you want to copy then `deepcopy(position(system))`.

File: src/atomsbase_interface.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    set_optimizable_mask(system::AbstractSystem, mask::AbstractVector{<:Bool})
+
+Sets the mask defining which coordinates of the system can be optimized. 
+The mask is a vector of booleans, specifying which of the atoms can be optimized.
+
+By default (when no mask is specified), all particles are assumed optimizable.
+
+"""
+function set_optimizable_mask(system::AbstractSystem, mask::AbstractVector{<:Bool})
+    particles = [Atom(atom; optimizable=m) for (atom, m) in zip(system.particles, mask)]
```
Comment (tjjarvinen) [inline src/atomsbase_interface.jl:50]: Here too

```julia
# Use this
zip(system, mask)

# Not this
zip(system.particles, mask)
```

File: src/atomsbase_interface.jl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+@doc raw"""
+    get_optimizable_mask(system::AbstractSystem) -> AbstractVector{<:Bool}
+
+Returns the optimizable mask of the system (see documentation for `set_optimizable_mask`.
+"""
+function get_optimizable_mask(system::AbstractSystem)
+    # If flag not set, the atom is considered to be optimizable.
+    [haskey(a, :optimizable) ? a[:optimizable] : true for a in system.particles]
```
Comment (tjjarvinen) [inline src/atomsbase_interface.jl:61]: Here too

```julia
# This
for a in system

# Not this
for a in system.particles
```

File: src/atomsbase_interface.jl [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+@doc raw"""
+    Clamp given atoms if the system. Clamped atoms are fixed and their positions 
+    will not be optimized. The atoms to be clamped should be given as a list of 
+    indies corresponding to their positions in system.particles.
+
+    """
+function clamp_atoms(system::AbstractSystem, clamped_indexes::Union{AbstractVector{<:Integer},Nothing})
+    mask = trues(length(system.particles))
```
Comment (tjjarvinen) [inline src/atomsbase_interface.jl:76]: Here too

```julia
# This
trues(length(system))

# Not this
trues(length(system.particles))
```

File: src/optimization.jl [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    By default we work in cartesian coordinaes.
+    Note that internally, when optimizing the cartesian positions, atomic units 
+    are used.
+
+"""
+function Optimization.OptimizationFunction(system::AbstractSystem, calculator; kwargs...)
+    mask = get_optimizable_mask(system) # mask is assumed not to change during optim.
+
+    f = function(x::AbstractVector{<:Real}, p)
+        x = 1u"bohr" .* x # Work in atomic units.
```
Comment (tjjarvinen) [inline src/optimization.jl:28]: You can use

```julia
x * u"bohr"
```

File: src/optimization.jl [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        new_system = update_optimizable_coordinates(system, x)
+        energy = AtomsCalculators.potential_energy(new_system, calculator; kwargs...)
+        austrip(energy)
+    end
+
+    g! = function(G::AbstractVector{<:Real}, x::AbstractVector{<:Real}, p)
+        x = 1u"bohr" .* x # Work in atomic units.
```
Comment (tjjarvinen) [inline src/optimization.jl:35]: Here too

```julia
x * u"bohr"
```

**Review by tjjarvinen — CHANGES_REQUESTED** [review]
My main issue is the use of `system.particles`. AtomsBase has an iterator over particles, so you can mostly just use `system` instead.

The second point is that `AbstractSystem` is an abstract type and both `FlexibleSystem` and `FastSystem` are it's subtypes. But only `FlexibleSystem` would work with the code.

File: src/atomsbase_interface.jl (hunk 1 above)
Comment (CedricTravelletti) [reply src/atomsbase_interface.jl]: I agree with your comment,  but I think that in the long run we should have an update constructor for `AbstractSystem` that updates positions by automatically dispatches to appropriate subtype constructor. 

Currently everything is converted to `FlexibleSystem`.

**Comment by CedricTravelletti** [comment]
Thaks @tjjarvinen for the helpful review. Fixes are now implemented. 
There is one remaining issue (see my comment about `AbstractSystem`.

File: Project.toml [hunk 8]
```diff
@@ -3,8 +3,26 @@ uuid = "673bf261-a53d-43b9-876f-d3c1fc8329c2"
...
+
 [compat]
-julia = "1"
+julia = "1.9"
+AtomsBase = "0.3"
+AtomsCalculators = "0.1"
+Optimization = "3.20"
+OptimizationOptimJL = "0.1"
+StaticArrays = "1.8"
+TestItemRunner = "0.2"
```
Comment (mfherbst) [inline Project.toml:23]: This is a test dependency, so it goes into the `[extras]` section and the `[targets]` section.

File: examples/Al_supercell.jl [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+using AtomsCalculators
+using Unitful
+using UnitfulAtomic
+using Random
+using OptimizationOptimJL
+
+using GeometryOptimization
+
+# Get PseudoDojo pseudopotential.
+psp_upf  = load_psp(artifact"pd_nc_sr_lda_standard_0.4.1_upf/Al.upf");
```
Comment (mfherbst) [inline examples/Al_supercell.jl:17]: You can inline that into the  build function

File: examples/Al_supercell.jl [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+# Starting position is a random perturbation of the equilibrium one.
+Random.seed!(1234)
+x0 = vcat(position(al_supercell)...)
+σ = 0.5u"angstrom"; x0_pert = x0 + σ * rand(Float64, size(x0))
+al_supercell = update_optimizable_coordinates(al_supercell, x0_pert)
+energy_pert = AtomsCalculators.potential_energy(al_supercell, calculator)
+
+@printf "Initial guess distance (norm) from true parameters %.3e bohrs.\n" austrip(norm(x0 - x0_pert))
+@printf "Initial regret %.3e.\n" energy_pert - energy_true
```
Comment (mfherbst) [inline examples/Al_supercell.jl:57]: No need to use printf for there. Just use print.

File: src/optimization.jl [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+        # NOTE: minus sign since forces are opposite to gradient.
+        G .= - austrip.(forces_concat)
+    end
+    OptimizationFunction(f; grad=g!)
+end
+
+function optimize_geometry(system::AbstractSystem, calculator; solver=Optim.LBFGS(), kwargs...)
+    # Use current system parameters as starting positions.
+    x0 = Vector(austrip.(get_optimizable_positions(system))) # Optim modifies x0 in-place, so need a mutable type.
```
Comment (mfherbst) [inline src/optimization.jl:49]: spacing (also in other places)

Comment (mfherbst) [reply src/optimization.jl:49]: Also the result of `austrip.` should be a Vector.

File: src/optimization.jl [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    f_opt = OptimizationFunction(system, calculator)
```
Comment (mfherbst) [inline src/optimization.jl:50]: I wonder if instead of creating a `OptimizationFunction` we should instead create an `OptimizationProblem`, because then we could also take more complex constraints into account more transparently later. Just a thought. Probably does not matter too much either way.

File: test/dummy_calculator.jl [hunk 13]
```diff
@@ -0,0 +1,17 @@
+# Create a dummy AtomsCalculator to test the geometry optimization interfcae.
+@testsetup module TestCalculators
+    using AtomsCalculators
+    using Unitful
+    using UnitfulAtomic
+    
+    struct DummyCalculator
+    end
```
Comment (mfherbst) [inline test/dummy_calculator.jl:8]: no line break

File: test/dummy_calculator.jl [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    
+    AtomsCalculators.@generate_interface function AtomsCalculators.potential_energy(system, calculator::DummyCalculator; kwargs...)
```
Comment (mfherbst) [inline test/dummy_calculator.jl:10]: line length

File: test/dummy_calculator.jl [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+        return 0.0u"eV"
+    end
+        
+    AtomsCalculators.@generate_interface function AtomsCalculators.forces(system, calculator::DummyCalculator; kwargs...)
+        return AtomsCalculators.zero_forces(system, calculator)
```
Comment (mfherbst) [inline test/dummy_calculator.jl:15]: returns

**Comment by cortner** [comment]
Why is there a type restriction `system::AbstractSystem` in the first place? Is there any place where we have a dispatch that requires this?

**Comment by mfherbst** [comment]
Probably not, but I think at least for optimize_geometry is useful to catch sanely programming errors where a user accidentally swaps the argument order, don't  you think?

**Comment by cortner** [comment]
My perspective is that abstract types are not supposed to be used to document, only to organize dispatch.

**Review by tjjarvinen — APPROVED** [review]

File: examples/Al_supercell.jl [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+optim_options = (f_tol=1e-6, iterations=6, show_trace=true)
+
+results = optimize_geometry(al_supercell, calculator; optim_options...)
```
Comment (cortner) [inline examples/Al_supercell.jl:61]: I am unsure about `optimize_geometry`. While it is true that the default would be to minimize energy, it is not the only meaning. Saddle point search and transition path search are also part of geometry optimisation. I think the canonical name is `minimize_energy` or something similar.  This is unambiguous.

File: src/atomsbase_interface.jl [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    update_positions(system, new_positions)
+end
+
+@doc raw"""
+Sets the mask defining which coordinates of the system can be optimized. 
+The mask is a vector of booleans, specifying which of the atoms can be optimized.
+
+By default (when no mask is specified), all particles are assumed optimizable.
+
+"""
```
Comment (cortner) [inline src/atomsbase_interface.jl:42]: I think this is too restrictive. There are also situations where you want to restrict specific coordinates. For example you can have atom positions be pinned to a surface but be allowed to slide on that surface. Maybe this doesn't have to be provided on the first implementation though. A note or an issue though so it can be done in the future.

File: src/atomsbase_interface.jl [hunk 18]
```diff
@@ -8,73 +8,66 @@
...
 Sets the mask defining which coordinates of the system can be optimized. 
 The mask is a vector of booleans, specifying which of the atoms can be optimized.
 
 By default (when no mask is specified), all particles are assumed optimizable.
 
 """
-function set_optimizable_mask(system::AbstractSystem, mask::AbstractVector{<:Bool})
-    particles = [Atom(atom; optimizable=m) for (atom, m) in zip(system.particles, mask)]
+function set_optimizable_mask(system, mask::AbstractVector{<:Bool})
+    particles = [Atom(atom; optimizable=m) for (atom, m) in zip(system, mask)]
```
[the remark this answers is not in the document]
Comment (cortner) [reply src/atomsbase_interface.jl:44]: I like clamped

File: src/atomsbase_interface.jl [hunk 19]
```diff
@@ -8,73 +8,66 @@
...
-
 Sets the mask defining which coordinates of the system can be optimized. 
 The mask is a vector of booleans, specifying which of the atoms can be optimized.
 
 By default (when no mask is specified), all particles are assumed optimizable.
 
 """
-function set_optimizable_mask(system::AbstractSystem, mask::AbstractVector{<:Bool})
-    particles = [Atom(atom; optimizable=m) for (atom, m) in zip(system.particles, mask)]
+function set_optimizable_mask(system, mask::AbstractVector{<:Bool})
```
[the remark this answers is not in the document]
Comment (cortner) [reply src/atomsbase_interface.jl:43]: should the function name have a !

File: src/atomsbase_interface.jl (hunk 17 above)
Comment (cortner) [reply src/atomsbase_interface.jl:42]: But the best way to deal with this is to make the mask completely general - e.g. a matrix, and have a simplified `clamp_atoms!` (as below).

File: src/atomsbase_interface.jl [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    collect(Iterators.flatten(system[mask, :position]))
+end
+
+@doc raw"""
+    Clamp given atoms if the system. Clamped atoms are fixed and their positions 
+    will not be optimized. The atoms to be clamped should be given as a list of 
+    indies corresponding to their positions in the system.
+
+    """
+function clamp_atoms(system, clamped_indexes::Union{AbstractVector{<:Integer},Nothing})
```
Comment (cortner) [inline src/atomsbase_interface.jl:69]: `clamp_atoms!`

File: src/optimization.jl [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+This is the reason for the cumbersome implementation of optimization with gradients.
+
+Note that by default all particles in the system are assumed optimizable.
+
+IMPORTANT: Note that we always work in cartesian coordinates.
+
+=#
+
+export optimize_geometry
```
Comment (cortner) [inline src/optimization.jl:15]: as discussed above.

**Review by cortner — CHANGES_REQUESTED** [review]
Am I right that the big missing gap is the optimization of the cell shape? This is an absolute must I think. Once we have that, and a more general mask, then the question how to organize DOFs becomes a little interesting. In JuLIP I had a `DofManager` that organized this for me.

**Comment by cortner** [comment]
I left some comments - we can iterate on what should be done now or can be postponed.

**Comment by CedricTravelletti** [comment]
Changes implemented. Ready for another round of review. 
There are two review suggestions that I am leaving pending: 

1. @tjjarvinen The `AbstractSystem`update constructor currently converts everything to `FlexibleSystem`. I believe that in the future, we should have an update constructor that can update the positions and return the correct system type (fast or flexible).
2. @cortner While I agree that there are different possible tasks when deforming the geometry (saddle point search, ...), usually, when someone talks about _geometry optimization_, by default they mean _minimize the energy_. That's also how other DFT packages use the term. I would thus suggest that we keep the `optimize_geometry` function, and have its default behaviour be to minimize the energy. Then, in the future, we might extend it and dispatch based on the _task_ (saddle point search or other).
What do you think?

**Comment by cortner** [comment]
> by default they mean minimize the energy

I somewhat agree, but I don't think that this excuses the ambiguity.

> Then, in the future, we might extend it and dispatch based on the task (saddle point search or other).

Yes we could define tasks or solvers whatever you want to call them and the run something more generic like `solve` or `optimize`
```julia
solver = EnergyMinimizer()
solver = DimerMethod() 
optimize(system, solver)
```
But then the debate just shifts to whether `EnergyMinimizer()` should be called `GeometryOptimizer()` (or whatever). So I don't think that will solve the problem. 

This isn't a big deal. We have a community package and the community can decide how to name things. If I'm outvoted then that's fine and as author you arguably have a bigger say. But my own proposal remains to be unambiguous and bearing in mind that the first three obvious geometry optimization tasks we will want to support are 
- minimize energy 
- find a saddle point
- compute the transition path
I suggest that we think of names that clearly specify these.

**Comment by tjjarvinen** [comment]
> 1. @tjjarvinen The `AbstractSystem`update constructor currently converts everything to `FlexibleSystem`. I believe that in the future, we should have an update constructor that can update the positions and return the correct system type (fast or flexible).

I am fine with this.

From my point all fine with the PR.

**Comment by antoine-levitt** [comment]
I guess it depends what is more obvious (and therefore can be left implied), if it's the energy or the geometry. Given the title of the package, @cortner 's suggestion seems reasonable to me.

(I guess when dealing with code or math, I typically think about optimizing the objective function, but when thinking more about the application I would think about optimizing the decision variables.)

(FWIW, when teaching students how to model concrete problems as mathematical optimization they seem to have more trouble figuring out the unknowns than the criterion, but I'm not sure what side, if any, this speaks in favor of)

**Comment by CedricTravelletti** [comment]
Given the above discussion, I agree with the suggestions and changed `optimize_geometry` to `minimize_energy`. 

I think all comments are addressed now.

**Comment by cortner** [comment]
I don't think so. Various comments have been unaddressed and not responded to.

**Comment by CedricTravelletti** [comment]
@cortner sorry I did not respond to the comments. I now added answers in the thread at the corresponding locations. 
To summarize what had been left out: 

- `clamp_atoms`: it was suggested that this should be `clamp_atoms!`. The problem is that this function relies on `AtomsBase`which currently does not provide in-place modifications. So I think we have to keep it as pure function. 
- Type restritction `system::AbstractSystem`: this has been removed. No type anymore. 
- Generalize clamping of atoms: I opened #2 to track the status of this. 
- Unit cell optimization: Working on this now. 

I hope this addresses what I had left out.

File: src/atomsbase_interface.jl [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+@doc raw"""
+    Clamp given atoms in the system. Clamped atoms are fixed and their positions 
+    will not be optimized. The atoms to be clamped should be given as a list of 
+    indices corresponding to their positions in the system.
+
+    """
+function clamp_atoms(system, clamped_indexes::Union{AbstractVector{<:Integer},Nothing})
+    clamped = falses(length(system))
+    clamped[clamped_indexes] .= true
+    particles = [Atom(atom; clamped=m) for (atom, m) in zip(system, clamped)]
```
Comment (rkurchin) [inline src/atomsbase_interface.jl:56]: An alternative way to do this (see also @mfherbst's [related comment](https://github.com/JuliaMolSim/GeometryOptimization.jl/pull/1/files#r1431521080)) that would be a bit more like what Optim.jl or Optimization.jl do would be to pass a separate `constraints` argument to the optimizer. I lean towards this approach because it allows for constraints that might not be a property of a single atom the way "being clamped" is – e.g. bond or symmetry constraints.

File: src/optimization.jl (hunk 12 above)
Comment (rkurchin) [reply src/optimization.jl:41]: I think this is probably a good idea, see my comment above – in SciML (Optimization.jl) I think they have both the `OptimizationFunction` (which is optional since you can pass a function directly as well) and the `OptimizationProblem` objects.

**Review by rkurchin — COMMENTED** [review]
Sorry for slow response here, I'm back online now – mostly LGTM, but I would strongly advocate for separating out constraints, see my comments for more detail.

File: src/atomsbase_interface.jl (hunk 22 above)
Comment (rkurchin) [reply src/atomsbase_interface.jl:56]: Just now noticed https://github.com/JuliaMolSim/GeometryOptimization.jl/issues/2 – I think this would also substantially simplify addressing that...

File: src/optimization.jl (hunk 12 above)
Comment (CedricTravelletti) [reply src/optimization.jl:41]: I agree that `OptimizationProblem` can be used in the future to implement constraints. Nevertheless, I would argue that we have to keep our definition of `OptimizationFunction` (as compared to passing directly a function to `OptimizationProblem`), since we need it because of the way we compute our gradients (so that some of the algorithms can compute both function value and gradient at the same time).

File: src/atomsbase_interface.jl (hunk 22 above)
Comment (CedricTravelletti) [reply src/atomsbase_interface.jl:56]: Hi Rachel. Thanks for the feedback. 
For the constraints, I suggest we defer implementating more fancy one to #2 and just work with fixed/clamped at the moment. 

Also, in terms of speed and dimensionality, I suspect that the approach we now have with `clamped` is useful, since it explicitly reduces the dimension of the problem before starting any optimization, whereas implementing this with a constraint in Optimization.jl will likely add some overhead. 

Long story short: I would advocate for keeping the clamping system for now, since it allows for efficient treatment of the basic situations. Then, once we have implemented #2 we can see if a separate function for the case of atoms clamping is still needed or not.

Comment (rkurchin) [reply src/atomsbase_interface.jl:90]: Hmm, I see what you're saying, and I definitely appreciate the imperative to start simple and only build in complications if they're needed, but given that we know that things like bond or symmetry constraints are real applications, I guess I just want to avoid putting something in that we'd then need to roll back, or creating confusing redundancies. I'm okay proceeding with the way it is now if others are okay with this, particularly @mfherbst as the DFTK person...

Comment (cortner) [reply src/atomsbase_interface.jl:90]: I would support keeping "clamps" as a special case. Even after we have more general constraints. In any case we want to make progress and this is a good start.

Comment (mfherbst) [reply src/atomsbase_interface.jl:90]: I agree. The current state is a good compromise for a first solution.

File: examples/TiAl_EmpiricalPotentials.jl [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+# Convert to AbstractSystem, so we have the `particles` attribute.
+particles = map(data) do atom
+    Atom(; pairs(atom)...)
+    end
+system = AbstractSystem(data; particles)
+
+solver = OptimizationOptimJL.LBFGS()
+optim_options = (f_tol=1e-8, g_tol=1e-8, iterations=10, show_trace=true)
+
+results = minimize_energy!(system, lj; solver=solver, optim_options...)
```
Comment (mfherbst) [inline examples/TiAl_EmpiricalPotentials.jl:25]:
```suggestion
results = minimize_energy!(system, lj; solver, optim_options...)
```

File: examples/TiAl_EmpiricalPotentials.jl [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+
+fname = joinpath(pkgdir(EmpiricalPotentials), "data/", "TiAl-1024.xyz")
+data = ExtXYZ.load(fname) |> FastSystem
+
+lj = LennardJones(-1.0u"meV", 3.1u"Å",  13, 13, 6.0u"Å")
+
+# Convert to AbstractSystem, so we have the `particles` attribute.
+particles = map(data) do atom
+    Atom(; pairs(atom)...)
+    end
```
Comment (mfherbst) [inline examples/TiAl_EmpiricalPotentials.jl:19]:
```suggestion
end
```

File: examples/TiAl_EmpiricalPotentials.jl [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+using EmpiricalPotentials
+using ExtXYZ
+using Unitful
+using UnitfulAtomic
+using OptimizationOptimJL
+
+using GeometryOptimization
+
+fname = joinpath(pkgdir(EmpiricalPotentials), "data/", "TiAl-1024.xyz")
+data = ExtXYZ.load(fname) |> FastSystem
```
Comment (mfherbst) [inline examples/TiAl_EmpiricalPotentials.jl:12]: Not a big fan of the `|>`, but ok ...

File: examples/H2_LJ.jl [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+using LinearAlgebra
+using EmpiricalPotentials
+using Unitful
+using UnitfulAtomic
+using OptimizationOptimJL
+using AtomsBase
+
+using GeometryOptimization
+
+
```
Comment (mfherbst) [inline examples/H2_LJ.jl:12]:
```suggestion
```

File: examples/H2.jl [hunk 27]
```diff
@@ -0,0 +1,34 @@
+#= Test Geometry Optimization on an aluminium supercell.
```
Comment (mfherbst) [inline examples/H2.jl:1]: @CedricTravelletti This is another prime example why too many comments can be bad :smile: ... there is nothing worse than wrong comments. If updating a comment is overlooked, it's likely not important enough to justify future maintenance cost.

File: src/atomsbase_interface.jl [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+    mask = not_clamped_mask(system)
+    Iterators.flatten(system[mask, :position])
+end
+
+@doc raw"""
+    Clamp given atoms in the system. Clamped atoms are fixed and their positions 
+    will not be optimized. The atoms to be clamped should be given as a list of 
+    indices corresponding to their positions in the system.
+
+    """
```
Comment (mfherbst) [inline src/atomsbase_interface.jl:86]:
```suggestion
Clamp given atoms in the system. Clamped atoms are fixed and their positions 
will not be optimized. The atoms to be clamped should be given as a list of 
indices corresponding to their positions in the system.
"""

File: src/optimization.jl [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+=#
+
+export minimize_energy!
+
+
+""" 
+    By default we work in cartesian coordinaes.
+    Note that internally, when optimizing the cartesian positions, atomic units 
+    are used.
+
```
Comment (mfherbst) [inline src/optimization.jl:13]:
```suggestion
```

File: src/optimization.jl [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+# IMPORTANT: Note that we always work in cartesian coordinates.
+=#
+
+export minimize_energy!
+
+
+""" 
+    By default we work in cartesian coordinaes.
+    Note that internally, when optimizing the cartesian positions, atomic units 
+    are used.
```
Comment (mfherbst) [inline src/optimization.jl:12]:
```suggestion
By default we work in cartesian coordinaes.
Note that internally, when optimizing the cartesian positions, atomic units 
are used.
```

**Review by mfherbst — COMMENTED** [review]
mostly formatting nits

File: src/optimization.jl (hunk 30 above)
Comment (CedricTravelletti) [reply src/optimization.jl]: Those formatting issues are now fixed. 
I agree on the comments stuff :) will be less verbose from now on.

File: src/atomsbase_interface.jl (hunk 22 above)
Comment (rkurchin) [reply src/atomsbase_interface.jl:89]: Alright, given comments above, I'm happy for this to be merged as-is. I merely reserve the right for an "I told you so" if we decide to change it later... 🤪

File: src/atomsbase_interface.jl (hunk 19 above)
[the remark this answers is not in the document]
Comment (cortner) [reply src/atomsbase_interface.jl]: see below - there is no in-place in AtomsBase

**Comment by cortner** [comment]
@CedricTravelletti , thanks for following up.  I want to take a quick look at the position and cell optimization before merging if that's ok?

**Comment by CedricTravelletti** [comment]
> @CedricTravelletti , thanks for following up. I want to take a quick look at the position and cell optimization before merging if that's ok?
> 
> I actually can't find examples or tests for optimizing a supercell with and without the cell shape included in the optimization? Can you point me to where I should look?

Dear Christoph 

With Michael we think that the part about cell shape optimization can go in a different PR, since there will probably be much more to discuss on that one. 
There is a preliminary version implemented in the `lattice` branch of the repository, and we are working on it this week to make it correct and benchmarked on realistic problems. 

So for now, I suggest that this PR only handles fixed cell optimization and that variable cell optimization is taken care of in a next PR mergin the `lattice` branch.

**Comment by cortner** [comment]
Ok.

**Comment by cortner** [comment]
In that case is it ready?

**Comment by rkurchin** [comment]
I'm happy for this to be merged! Perhaps we should open an issue regarding the cell shape optimization just to keep visible that that's a to-do item?

**Comment by CedricTravelletti** [comment]
Yes, the non unit-cell optimization part is ready.
